src/services/rfqm_service.ts

```diff
@@ -0,0 +1,136 @@
+    ) {}
+
+    /**
+     * Fetch the best indicative quote available. Returns null if no valid quotes found
+     */
+    public async fetchIndicativeQuoteAsync(
+        params: FetchIndicativeQuoteParams,
+    ): Promise<FetchIndicativeQuoteResponse | null> {
+        // Extract params
+        const {
+            sellAmount,
+            buyAmount,
+            sellToken: takerToken,
+            buyToken: makerToken,
+            sellTokenDecimals: takerTokenDecimals,
+            buyTokenDecimals: makerTokenDecimals,
+            apiKey,
+        } = params;
+
+        // Quote Requestor specific params
+        const isSelling = sellAmount !== undefined;
+        const marketOperation = isSelling ? MarketOperation.Sell : MarketOperation.Buy;
+        const assetFillAmount = isSelling ? sellAmount! : buyAmount!;
+
+        // Prepare gas estimate
+        const gas = await this._protocolFeeUtils.getGasPriceEstimationOrThrowAsync();
+
+        // Fetch quotes
+        const opts = {
+            ...RFQM_DEFAULT_OPTS,
+            apiKey,
+            intentOnFilling: false,
+            isIndicative: true,
+        };
+        const indicativeQuotes = await this._quoteRequestor.requestRfqmIndicativeQuotesAsync(
+            makerToken,
+            takerToken,
+            assetFillAmount,
+            marketOperation,
+            undefined,
+            opts,
+        );
+
+        // Filter out quotes that:
+        // - are for the wrong pair
+        // - cannot fill 100 % of the requested amount
+        // - expire in less than 3 minutes
+        //
+        // And sort by best price
+        const now = new BigNumber(Date.now());
+        const expirationCutoff = now.plus(RFQM_MINUMUM_EXPIRY_DURATION_MS).div(ONE_SECOND_MS);
+        const sortedQuotes = indicativeQuotes
+            .filter((q) => q.takerToken === takerToken && q.makerToken === makerToken)
+            .filter((q) => {
+                const requestedAmount = isSelling ? q.takerAmount : q.makerAmount;
+                return requestedAmount.gte(assetFillAmount);
+            })
+            .filter((q) => q.expiry.gte(expirationCutoff))
+            .sort((a, b) => {
+                // Want the most amount of maker tokens for each taker token
+                const aPrice = a.makerAmount.div(a.takerAmount);
+                const bPrice = b.makerAmount.div(b.takerAmount);
+                return bPrice.minus(aPrice).toNumber();
+            });
+
+        // No quotes found
+        if (sortedQuotes.length === 0) {
+            return null;
+        }
+
+        // Get the best quote
+        const bestQuote = sortedQuotes[0];
+
+        // Prepare the price
+        const makerAmountInUnit = Web3Wrapper.toUnitAmount(bestQuote.makerAmount, makerTokenDecimals);
+        const takerAmountInUnit = Web3Wrapper.toUnitAmount(bestQuote.takerAmount, takerTokenDecimals);
+        const price = makerAmountInUnit.div(takerAmountInUnit);
```

Comment:
Just checking - is price always maker over taker or could it be the other way round? How do other services (like swap service) do it?

Reply:
Good call! this is not how swap service currently handles it - will update to take into account selling vs buying